### ISSUE TITLE
Simple add on Lists failing in 1.1.0+

### ISSUE THREAD

**Issue by PartyArk** [aspnet/JsonPatch#48]
Working on 1.0.0, but failing on 1.1.0 and 1.2.0-preview1-22759

Here I'm attempting to add a new entry on List named Mylist which already has two entries (at indexes zero and one)

`[{"op":"add","path":"/mylist/2","value":{"myname":"foo","age":33}}]`

Response:
JsonPatchException: The index value provided by path segment '2' is out of bounds of the array size.

I'm guessing it's incorrectly looking at the length of the list prior to adding. With a dash at the end instead of the new index, this **will** work:

`[{"op":"add","path":"/mylist/-","value":{"myname":"foo","age":33}}]`

**Comment by PartyArk** [comment]
The specification says:

   o  An element to add to an existing array - whereupon the supplied
      value is added to the array at the indicated location.  Any
      elements at or above the specified index are shifted one position
      to the right.  The specified index MUST NOT be greater than the
      number of elements in the array.  If the "-" character is used to
      index the end of the array (see [RFC6901]), this has the effect of
      appending the value to the array.

So if we have a list with 2 items, an _index_ of 2 is conformant because it it _not greater_ than the number of elements in the array, and also it implies the same thing as "-" ie. it should add to the end of the array. 

We use the popular fast-json js library, where this same issue has been discussed - https://github.com/Starcounter-Jack/JSON-Patch/issues/78

**Comment by PartyArk** [comment]
I've looked at the code - the reason this worked previouly is that using IList.Insert, if the index is the length of the list, it was appended on the end. See ObjectAdapter.cs 1.0.0.

For 1.1.0+ there is a new method TryGetPositionInfo which creates an error if the position is less than the length of the list.Count. This is incorrect behaviour. We need to return PositionType.EndOfList in this scenario, by introducing an additional check at line 258

```
      if (position >= 0 && position == list.Count)
             {
                positionInfo = new PositionInfo(PositionType.EndOfList, -1);
                errorMessage = null;
                return true;
            }
```

**Comment by Eilon** [comment]
Indeed, this sounds like a regression. We will investigate a fix. Thanks!

**Comment by kichalla** [comment]
This was my bad...not sure how my missed the `number of elements` part..

**Comment by Eilon** [comment]
This patch bug is **approved**. Please use the normal code review process w/ a PR and make sure the fix is in the correct branch, then close the bug and mark it as done.

**Comment by Eilon** [comment]
@kichalla please make sure the fix is in 2.0.0 as well.

**Comment by Eilon** [comment]
@kichalla I opened https://github.com/aspnet/JsonPatch/issues/54 to track making sure the fix is also in 2.0.0.
